pkg/storage/mvcc.go

```diff
@@ -4256,6 +4301,221 @@ func MVCCGarbageCollect(
+		KeyTypes:     IterKeyTypePointsOnly,
+		StartKey:     rangeStartKey,
+		EndKey:       rangeEndKey,
+		EndTime:      highestTS,
+		IntentPolicy: MVCCIncrementalIterIntentPolicyEmit,
+	})
+	defer ptIter.Close()
+
+	for _, gcKey := range rks {
+		iter.SeekGE(MVCCKey{Key: gcKey.StartKey})
+
+		for ; ; iter.Next() {
+			if ok, err := iter.Valid(); err != nil {
+				return err
+			} else if !ok {
+				break
+			}
+
+			bounds := iter.RangeBounds()
+			startKey := bounds.Key
+			// Termination condition for the case where GC'd range ends before
+			// iterator end.
+			if startKey.Compare(gcKey.EndKey) >= 0 {
+				break
+			}
+
+			// Check if we have a partial overlap between fragment and requested
+			// GCd range. This shouldn't happen normally, but we can accommodate for
+			// the case correctly instead of just bailing out.
+			trimLeft, trimRight := false, false
+			candidateStartKey := bounds.Key
+			if gcKey.StartKey.Compare(candidateStartKey) > 0 {
+				candidateStartKey = gcKey.StartKey
+				trimLeft = true
+			}
+			candidateEndKey := bounds.EndKey
+			if gcKey.EndKey.Compare(candidateEndKey) < 0 {
+				candidateEndKey = gcKey.EndKey
+				trimRight = true
+			}
+
+			removedRange := MVCCRangeKeyValue{
+				RangeKey: MVCCRangeKey{
+					StartKey: candidateStartKey,
+					EndKey:   candidateEndKey,
+				},
+			}
+			for i, rkv := range iter.RangeKeys() {
+				removedRange.RangeKey.Timestamp = rkv.RangeKey.Timestamp
+				removedRange.Value = rkv.Value
+				remove := rkv.RangeKey.Timestamp.LessEq(gcKey.Timestamp)
+				if remove {
+					if err := rw.ExperimentalClearMVCCRangeKey(removedRange.RangeKey); err != nil {
+						return err
+					}
+				}
+				if ms != nil {
+					ms.Add(updateStatsOnRangeTombstoneGC(removedRange, trimLeft, trimRight, i == 0, remove))
+				}
+			}
+
+			// Verify that there are no remaining data under the deleted range using
+			// time bound iterator.
+			ptIter.SeekGE(MVCCKey{Key: candidateStartKey})
+			for ; ; ptIter.Next() {
+				if ok, err := ptIter.Valid(); err != nil {
+					return err
+				} else if !ok {
+					break
+				}
+				pointKey := ptIter.UnsafeKey()
+				if pointKey.Key.Compare(candidateEndKey) >= 0 {
+					break
+				}
+				if !pointKey.IsValue() {
+					var meta enginepb.MVCCMetadata
+					if err := protoutil.Unmarshal(ptIter.UnsafeValue(), &meta); err != nil {
+						return err
+					}
+					if meta.Timestamp.Less(gcKey.Timestamp.ToLegacyTimestamp()) {
+						return errors.Errorf("attempt to delete range tombstone %q hiding intent at %q",
+							gcKey, pointKey)
+					}
+				} else if pointKey.Timestamp.Less(gcKey.Timestamp) {
+					val, err := DecodeMVCCValue(ptIter.UnsafeValue())
+					if err != nil {
+						return err
+					}
+					if !val.IsTombstone() {
+						return errors.Errorf("attempt to delete range tombstone %q hiding value at %q",
+							gcKey, pointKey)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func updateStatsOnRangeTombstoneGC(
```

Comment:
This logic doesn't handle range tombstone merges. Consider this case:

```
5  o---o---o---o
4  o---o---o---o
3
2      o---o
1
   a   b   c   d
```

If we GC below 3, then the removal of `[b-c)@2` will cause the above fragments to merge into `[a-d)@4` and `[a-d)@5`, reducing e.g. `RangeKeyCount` from 3 to 1.

Why didn't the randomized GC tests catch this?